webextension/background.js

```diff
@@ -1,3 +1,63 @@
+browser.contextMenus.create({
+  id: "open-in-this-container",
+  title: "Always Open in This Container",
+  contexts: ["all"],
+});
+
+function getUserContextIdFromCookieStore(cookieStore) {
+  const container = cookieStore.replace("firefox-container-", "");
+  if (container !== cookieStore) {
+    return container;
+  }
+  return false;
+}
+
+const storagePrefix = "originMap@@_";
+const storageArea = browser.storage.local;
+
+browser.contextMenus.onClicked.addListener((info, tab) => {
+  const pageUrl = new window.URL(info.pageUrl);
+  const userContextId = getUserContextIdFromCookieStore(tab.cookieStoreId);
+  // Mapping ${pageUrl.host} to ${userContextId}
+  if (userContextId) {
+    const originStoreKey = `${storagePrefix}${pageUrl.host}`;
+    storageArea.set({[originStoreKey]: userContextId});
+  }
+});
+
+browser.webRequest.onBeforeRequest.addListener((options) => {
+  if (options.tabId === -1) {
+    return;
+  }
+  const pageUrl = new window.URL(options.url);
+  const originStoreKey = `${storagePrefix}${pageUrl.host}`;
+   return Promise.all([
+       browser.tabs.get(options.tabId),
+       storageArea.get([originStoreKey])
+     ]).then(([tab, storage]) => {
+     const userContextId = getUserContextIdFromCookieStore(tab.cookieStoreId);
+     if (!(originStoreKey in storage)) {
+       return;
+     }
+     const mapUserContextId = storage[originStoreKey];
+     if (userContextId === mapUserContextId) {
+       return {};
+     }
+     browser.tabs.create({url: options.url, cookieStoreId: `firefox-container-${mapUserContextId}`});
+     // If the user just opened the tab, we can auto close it
+     if (tab.url == "about:newtab" || tab.url === "about:home") {
+       browser.tabs.remove(tab.id);
+     }
+     return {
+      cancel: true,
+     };
+     return {};
+  }).catch((e) => {
+    console.log('err', e);
+    return {};
+  });
+},{urls: ["<all_urls>"], types: ["main_frame"]}, ["blocking"]);
```

Comment:
Ah, the `main_frame` `RequestFilter` here takes care of my concern about processing every request. 👍